### PR TITLE
in which inferable outlives-requirements are linted

### LINKED ISSUE #52042: edition lint: unnecessary `T: 'a` annotations
We might want to wait until https://github.com/rust-lang/rust/issues/51858 is implemented before landing (not sure), but we can begin implementation.

### PR DESCRIPTION
**Pull request by zackmdavis**
RFC 2093 (tracking issue #44493) lets us leave off these
commonsensically inferable `T: 'a` outlives requirements. (A separate
feature-gate was split off for the case of 'static lifetimes, for
which questions still remain.) Detecting these was requested as an
idioms-2018 lint.

Resolves #52042, an item under the fabulous metaïssue #52047.

It's plausible that this shouldn't land until after `infer_outlives_requirements` has been stabilized ([final comment period started](https://github.com/rust-lang/rust/issues/44493#issuecomment-408852946) 4 days ago), but I think there's also a strong case to not-wait in order to maximize the time that [Edition Preview 2](https://internals.rust-lang.org/t/rust-2018-release-schedule-and-extended-beta/8076) users have to kick at it. (It's allow by default, so there's no impact unless you explicitly turn it or the rust-2018-idioms group up to `warn` or higher.)

Questions—

 * Is `explicit-outlives-requirements` a good name? (I chose it as an [RFC 344](https://github.com/rust-lang/rfcs/blob/master/text/0344-conventions-galore.md#lints)-compliant "inversion" of the feature-gate name, `infer_outlives_requirements`, but I could imagine someone arguing that the word `struct` should be part of the name somewhere, for specificity.)

 * Are there any false-positives or false-negatives? @nikomatsakis [said that](https://github.com/rust-lang/rust/issues/52042#issuecomment-406409795) getting this right would be "fairly hard", which makes me nervous that I'm missing something. The UI test in the initial submission of this pull request just exercises the examples [given in the Edition Guide](https://rust-lang-nursery.github.io/edition-guide/2018/transitioning/ownership-and-lifetimes/struct-inference.html).

infer_outlints

r? @alexcrichton

### CODE REVIEW

**Comment by zackmdavis** [comment]
(CI failure is tidy/forgetting-to-bless-after-tidy-fix triviality, can fix tonight. Should also test multi-predicate where clauses, for which I expect this initial implementation to not get the span/suggestion right.)

File: src/librustc_lint/builtin.rs
Comment (alexcrichton) [inline src/librustc_lint/builtin.rs]: Should the `is_static` condition be dropped here as well? Ideally we wouldn't fire lints unless the fix works becaue the crate already has the feature enabled

**Comment by alexcrichton** [comment]
Looks good to me, thanks!

I think we'll also want a test to ensure the lint doesn't fire when the feature isn't enabled, but otherwise r=me

File: src/librustc_lint/builtin.rs
Comment (zackmdavis) [reply src/librustc_lint/builtin.rs]: `infer_static_outlives_requirements` (note, not the same as `infer_outlives_requirements`) is a [new feature gate](https://github.com/rust-lang/rust/issues/44493#issuecomment-407848932) specifically for the special case of static lifetimes that there [were concerns](https://github.com/rust-lang/rust/issues/44493#issuecomment-407846046) about.

I was reluctant to gate the entire lint on `infer_outlives_requirements` since we're so close to stabilizing it, but I guess it's not very much work (and not much work to undo when we stabilize, even if that's only a few weeks from now) ...

**Comment by zackmdavis** [comment]
Update—

> Should also test multi-predicate where clauses

It turns out that getting the spans right in the presence of multiple bounds and multi-predicate where clauses is actually pretty involved. I'm _pretty close_ to getting this right in full generality (see all the new code in the UI test), but not quite there yet. (And out of time for today.)

File: src/librustc_lint/builtin.rs
Comment (alexcrichton) [reply src/librustc_lint/builtin.rs]: Ah yeah due to the way `cargo fix` works it's best to ignore this lint unless the feature is turned on as otherwise lots of false positives get emitted

**Comment by zackmdavis** [comment]
Well, that was [intense](https://gist.github.com/zackmdavis/839eb750fe3bd8152b0d5ec6faad87ea), but I think the work here is done!—but unfortunately, the run-rustfix test is blocked on rust-lang-nursery/rustfix#141 (which I can also work on). So this should be ready to merge after that gets fixed and _either_ the feature gets stabilized (which @toidiu [is going to do](https://github.com/rust-lang/rust/issues/44493#issuecomment-411788473)) or we add a trivial don't-lint-if-feature-gated conditional here.

ETA the [day after RustConf](https://www.eventbrite.com/e/rust-hack-day-post-rustconf-2018-tickets-47903658131)?

File: src/librustc_lint/builtin.rs
Comment (nikomatsakis) [inline src/librustc_lint/builtin.rs]: So this just removes **all** outlives requirements -- but I think there are still cases where outlives declarations are required, though they are somewhat obscure. You can find an example in the ["where outlives requirements are still required" section of the RFC](https://rust-lang.github.io/rfcs/2093-infer-outlives.html#where-explicit-annotations-would-still-be-required). 

Probably the *right* way to do this would be to actually run the [`inferred_outlives_of`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.TyCtxt.html#method.inferred_outlives_of) query and check whether this bound appears in the result. If so, it can be removed.

Comment (zackmdavis) [reply src/librustc_lint/builtin.rs]: Thanks for the pointer! :sparkling_heart:

**Comment by zackmdavis** [comment]
@alexcrichton cc @nikomatsakis :checkered_flag: 

We [now actually use](https://github.com/rust-lang/rust/blob/27b2237bde8f0f6fdaeb25cf57e69b07da798593/src/librustc_lint/builtin.rs#L1971-L2010) the `inferred_outlives_of` query to only lint if the bound really is inferrable. A [new compile-pass UI test](https://github.com/rust-lang/rust/blob/27b2237bde8f0f6fdaeb25cf57e69b07da798593/src/test/ui/rust-2018/edition-lint-uninferable-outlives.rs) with empty expected output verifies that we don't lint the example of a non-inferable outlives-bound from the RFC (whereas previous revisions of this PR did erroneously lint). Some of the UI test examples for which the lint emits multiple spans were split into a separate file because rust-lang-nursery/rustfix#141 prevents the suggestions from being verified with the `run-rustfix` directive. (I had previously thought I could address rust-lang-nursery/rustfix#141 before we try to land the present PR, but now I think there are some serious design concerns there which I will elaborate on in a forthcoming issue.)

**Comment by alexcrichton** [comment]
Nice! This looks generally good to me (good tests, good code/comments/etc), but I'm not looking too much at the particulars as I'm not really overly familiar with them. @nikomatsakis do you think you'll have a chance to review this more closely? If not I can always r+ :)

**Comment by TimNN** [comment]
Ping from triage @nikomatsakis: Some feedback has been requested from you for this PR.

**Comment by TimNN** [comment]
Ping from triage @nikomatsakis: Some feedback has been requested from you for this PR.

**Comment by zackmdavis** [comment]
(This is going to need a rebase on #53793 anyway.)

**Comment by zackmdavis** [comment]
@nikomatsakis @alexcrichton rebased (now that infer-outlives is stablized) and removed the feature gates :checkered_flag:

File: src/librustc_lint/builtin.rs
Comment (nikomatsakis) [inline src/librustc_lint/builtin.rs]: You can also have a scenario like

```rust
struct Foo<'a, 'b: 'a> {
    x: &'a &'b u32
}
```

It'd be nice to remove those as well!

Comment (nikomatsakis) [reply src/librustc_lint/builtin.rs]: Another example (much more obscure) is with projections:

```rust
struct Foo<'a, T: Iterator> where T::Item: 'a {
    item: T::Item,
}
```

**Review by nikomatsakis — APPROVED** [review]
Hacky but beautiful. I'm game to land this as is, but as I noted, it doesn't actually cover the full range of cases. What do you want to do about the remainder?

**Comment by nikomatsakis** [comment]
@zackmdavis so sorry for the delay, this slipped my radar for a while :(

r=me if you want to land as is, in which case we should file a follow-up issue for the others

**Comment by zackmdavis** [comment]
Rebased to remove the commit adding `multipart_suggestion_with_applicability`, which was [independently added in #54241](https://github.com/rust-lang/rust/commit/2f5cb6dbdc7aff9d4c5991c9eea78b5d9daed815/).

Filed #54630 for the false-negatives.

@bors r=nikomatsakis

**Pull request merged**
